### ISSUE TITLE
Numerical operator does not work on re-written values

### ISSUE THREAD

**Issue by Wernfried** [balabit/syslog-ng#1105]
Hi,

In my syslog messages I have an IP-Address and I like to split destination based on IP-Ranges, i.e. value of last octet.

This is relevant part of my config file:

```
log { 
	source(s_firewall);
	filter { level(notice); }; 
	parser { kv-parser(prefix(".IPMSG.")); };
	filter { "${.IPMSG.type}" eq "traffic" }; 

	rewrite { 
		set("${.IPMSG.srcip}", value("LAST_OCTET") );
		subst("[[:digit:]]+.[[:digit:]]+.[[:digit:]]+.", "", value("LAST_OCTET") );
	};
	
	log { filter { "${LAST_OCTET}" >= "224" }; destination(d_SQLite_224); flags(final); };
 	log { filter { "${LAST_OCTET}" >= "192" }; destination(d_SQLite_192); flags(final); };
 	log { filter { "${LAST_OCTET}" >= "160" }; destination(d_SQLite_160); flags(final); };
 	log { filter { "${LAST_OCTET}" >= "128" }; destination(d_SQLite_128); flags(final); };
 	log { filter { "${LAST_OCTET}" >= "96" }; destination(d_SQLite_96); flags(final); };
 	log { filter { "${LAST_OCTET}" >= "64" }; destination(d_SQLite_64); flags(final); };
 	log { filter { "${LAST_OCTET}" >= "32" }; destination(d_SQLite_32); flags(final); };
 	log { filter { "${LAST_OCTET}" >= "0" }; destination(d_SQLite_0); flags(final); };

	flags(flow-control);
};

```
I tried different ways, e.g. `filter { "${LAST_OCTET}" >= value("224") };` or `set($+ value("LAST_OCTET") value("0") );` but none of them worked.

All messages are going to destination "d_SQLite_0" or d_"SQLite_224", so it looks like syslog-ng does a string comparison rather than numeric comparison. Is there a way to achieve numeric comparison? Otherwise I have to write 255 times

```
log { filter { "${LAST_OCTET}" eq "0" }; destination(d_SQLite_0); flags(final); };
log { filter { "${LAST_OCTET}" eq "1" }; destination(d_SQLite_0); flags(final); };
log { filter { "${LAST_OCTET}" eq "2" }; destination(d_SQLite_0); flags(final); };
log { filter { "${LAST_OCTET}" eq "3" }; destination(d_SQLite_0); flags(final); };
...
log { filter { "${LAST_OCTET}" eq "255" }; destination(d_SQLite_224); flags(final); };

```

Best Regards
Wernfried

**Comment by bazsi** [comment]
There was an issue with numeric vs string comparisons that was fixed in 3.8
which you have ran into.

The fix is here:

https://github.com/balabit/syslog-ng/pull/774

Hi,

In my syslog messages I have an IP-Address and I like to split destination
based on IP-Ranges, i.e. value of last octet.

This is relevant part of my config file:

log {
    source(s_firewall);
    filter { level(notice); };
    parser { kv-parser(prefix(".IPMSG.")); };
    filter { "${.IPMSG.type}" eq "traffic" };

    rewrite {
        set("${.IPMSG.srcip}", value("LAST_OCTET") );
        subst("[[:digit:]]+.[[:digit:]]+.[[:digit:]]+.", "",
value("LAST_OCTET") );
    };

    log { filter { "${LAST_OCTET}" >= "224" };
destination(d_SQLite_224); flags(final); };
    log { filter { "${LAST_OCTET}" >= "192" };
destination(d_SQLite_192); flags(final); };
    log { filter { "${LAST_OCTET}" >= "160" };
destination(d_SQLite_160); flags(final); };
    log { filter { "${LAST_OCTET}" >= "128" };
destination(d_SQLite_128); flags(final); };
    log { filter { "${LAST_OCTET}" >= "96" };
destination(d_SQLite_96); flags(final); };
    log { filter { "${LAST_OCTET}" >= "64" };
destination(d_SQLite_64); flags(final); };
    log { filter { "${LAST_OCTET}" >= "32" };
destination(d_SQLite_32); flags(final); };
    log { filter { "${LAST_OCTET}" >= "0" }; destination(d_SQLite_0);
flags(final); };

    flags(flow-control);
};


All messages are going to destination "d_SQLite_0" or d_"SQLite_224", so it
looks like syslog-ng does a string comparison rather than numeric
comparison. Is there a way to achieve numeric comparison? Otherwise I have
to write 255 times

log { filter { "${LAST_OCTET}" eq "0" }; destination(d_SQLite_0);
flags(final); };
log { filter { "${LAST_OCTET}" eq "1" }; destination(d_SQLite_0);
flags(final); };
log { filter { "${LAST_OCTET}" eq "2" }; destination(d_SQLite_0);
flags(final); };
log { filter { "${LAST_OCTET}" eq "3" }; destination(d_SQLite_0);
flags(final); };
...
log { filter { "${LAST_OCTET}" eq "255" }; destination(d_SQLite_224);
flags(final); };


Best Regards
Wernfried

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/balabit/syslog-ng/issues/1105>, or mute the thread
<https://github.com/notifications/unsubscribe/AArldkh_tD2fShTVIIbYF6gZShLHSY53ks5qQmYmgaJpZM4JBFPN>
.

**Comment by Wernfried** [comment]
Thak you, I will try.

By involving some little math I found also a workaround:

```
log { 
    source(s_firewall);
    filter { level(notice); }; 
    parser { kv-parser(prefix(".IPMSG.")); };
    filter { "${.IPMSG.type}" eq "traffic" }; 

    rewrite { 
		set("${.IPMSG.srcip}", value("LAST_OCTET") );
		subst("[[:digit:]]+.[[:digit:]]+.[[:digit:]]+.", "", value("LAST_OCTET") );
		set("$(/ $(- ${LAST_OCTET} $(% ${LAST_OCTET} '32')) '32')", value("IP_RANGE") );		
    };

	log { filter { "${IP_RANGE}" eq "0" }; destination(d_SQLite_0); flags(final); };
	log { filter { "${IP_RANGE}" eq "1" }; destination(d_SQLite_32); flags(final); };
	log { filter { "${IP_RANGE}" eq "2" }; destination(d_SQLite_64); flags(final); };
	log { filter { "${IP_RANGE}" eq "3" }; destination(d_SQLite_96); flags(final); };
	...
	
};

```
Best Regards
